src/routes/JoinGroup/JoinGroup.js

```diff
@@ -1,65 +1,74 @@
-import React, { useEffect, useState } from 'react'
+import React, { useEffect } from 'react'
 import { useDispatch, useSelector } from 'react-redux'
-import { Redirect } from 'react-router-dom'
+import { useHistory } from 'react-router-dom'
+import { every, isEmpty } from 'lodash/fp'
 import { groupUrl } from 'util/navigation'
+import fetchForGroup from 'store/actions/fetchForGroup'
 import getQuerystringParam from 'store/selectors/getQuerystringParam'
 import getRouteParam from 'store/selectors/getRouteParam'
-import getSignupState, { SignupState } from 'store/selectors/getSignupState'
+import { getSignupComplete } from 'store/selectors/getSignupState'
 import useInvitation from 'store/actions/useInvitation'
 import checkInvitation from 'store/actions/checkInvitation'
 import Loading from 'components/Loading'
-import setReturnToURL from 'store/actions/setReturnToURL'
+import setReturnToPath from 'store/actions/setReturnToPath'
 
 export const SIGNUP_PATH = '/signup'
 export const EXPIRED_INVITE_PATH = '/invite-expired'
 
 export default function JoinGroup (props) {
+  const history = useHistory()
   const dispatch = useDispatch()
-  const signupState = useSelector(getSignupState)
-  const [loading, setLoading] = useState(true)
-  const [isValidInvite, setIsValidInvite] = useState()
-  const [groupSlug, setGroupSlug] = useState()
-  const signupComplete = signupState === SignupState.Complete
-  const invitationTokenAndCode = {
-    invitationToken: getQuerystringParam('token', null, props),
-    accessCode: getRouteParam('accessCode', null, props)
-  }
+  const signupComplete = useSelector(getSignupComplete)
 
   useEffect(() => {
     (async function () {
-      setLoading(true)
       try {
+        const invitationTokenAndCode = {
+          invitationToken: getQuerystringParam('token', null, props),
+          accessCode: getRouteParam('accessCode', null, props)
+        }
+
+        if (every(isEmpty, invitationTokenAndCode)) {
+          throw new Error('Please provide either a `token` query string parameter or `accessCode` route param')
+        }
+
         if (signupComplete) {
           const useInvitationResult = await dispatch(useInvitation(invitationTokenAndCode))
           const newMembership = useInvitationResult?.payload?.getData()?.membership
-          setGroupSlug(newMembership?.group?.slug)
+          const groupSlug = newMembership?.group?.slug
+
+          if (groupSlug) {
+            /*
+              `AuthLayoutRouter` will already try and fetch this group due to the
+              `/groups/:groupSlug/join/<token>` route matching `:groupSlug` before the
+              group has been joined (unauthorized), this could be fixed and this extra
```

Comment:
file a ticket for this?

Reply:
join group is definitely slower than it needs to be because it loads the entire me query when it doesnt need to, seems like this is fixed here which is great!

Reply:
Fixed this.